QHyper/problems/community_detection.py

```diff
@@ -0,0 +1,193 @@
+            )
+
+        self.cases: int = N_communities
+        self.variables: tuple[
+            sympy.Symbol
+        ] = self._encode_discretes_to_one_hots()
+        self._set_objective_function()
+        self._set_one_hot_constraints(N_communities)
+
+    def _encode_discrete_to_one_hot(
+        self, discrete_variable: sympy.Symbol, case_value: int
+    ) -> sympy.Symbol:
+        discrete_id = int(str(discrete_variable)[len("x") :])
+        id = discrete_id * self.cases + case_value
+        return sympy.symbols(f"s{id}")
+
+    def _encode_discretes_to_one_hots(self) -> tuple[sympy.Symbol]:
+        one_hots: tuple[sympy.Symbol] = sympy.symbols(
+            " ".join(
+                [
+                    str(self._encode_discrete_to_one_hot(var, case_val))
+                    for var in self._get_discrete_variable_representation()
+                    for case_val in range(self.cases)
+                ]
+            )
+        )
+        return one_hots
+
+    def iter_variables_cases(self) -> Iterable[Tuple[sympy.Symbol, ...]]:
+        """s -> (s0,s1,s2,...sn-1), (sn,sn+1,sn+2,...s2n-1), ..."""
+        return zip(*[iter(self.variables)] * self.cases)
+
+    def _get_discrete_variable_representation(
+        self,
+    ) -> tuple[sympy.Symbol] | Any:
+        return sympy.symbols(
+            " ".join([f"x{i}" for i in range(len(self.G.nodes))])
+        )
+
+    def _set_objective_function(self) -> None:
+        equation: dict[VARIABLES, float] = {}
+        for i in self.G.nodes:
+            for j in range(i + 1, len(self.G.nodes)):
+                for case_val in range(self.cases):
+                    x_i, x_j = sympy.symbols(f"x{i}"), sympy.symbols(f"x{j}")
+                    s_i = str(self._encode_discrete_to_one_hot(x_i, case_val))
+                    s_j = str(self._encode_discrete_to_one_hot(x_j, case_val))
+                    equation[(s_i, s_j)] = self.B[i, j]
+
+        equation = {key: -1 * val for key, val in equation.items()}
+
+        self.objective_function = Expression(equation)
+
+    def _set_one_hot_constraints(self, N_communities: int) -> None:
+        ONE_HOT_CONST = -1
+        self.constraints: list[Expression] = []
+        # In the case of 1-to-1 mapping between discrete
+        # and binary variable values no one-hot constraints
+        if N_communities == ONE_HOT_CONST * -1:
+            return
+
+        dummies: Iterable[Tuple[sympy.Symbol, ...]]
+        for dummies in self.iter_variables_cases():
+            expression: Expr = cast(Expr, 0)
+            dummy: sympy.Symbol
+            for dummy in dummies:
+                expression += dummy
+            expression += ONE_HOT_CONST
+            self.constraints.append(Expression(expression))
+
+    def decode_solution(self, solution: dict) -> dict:
+        ONE_HOT_VALUE = 1.0
+        decoded_solution: dict = {}
+
+        for variable, value in solution.items():
+            _, id = variable[0], int(variable[len("s") :])
```

Comment:
That also should work right? xd And again len("s") is always 1
```suggestion
            id = int(variable[len("s") :])
```